docs/src/proposals/off-chain-message-signing.md

```diff
@@ -0,0 +1,134 @@
+
+#### Application domain
+
+A 32-byte array identifying the application requesting off-chain message signing.
+This may be any arbitrary bytes. For instance the on-chain address of a program,
+DAO instance, Candy Machine, etc.
+
+This field **SHOULD** be displayed to users as a base58-encoded ASCII string rather
+than interpretted otherwise.
+
+#### Message Format
+
+Version `0` headers specify three message formats allowing for trade-offs between
+compatibility and composition of messages.
+
+| ID  | Encoding              | Maximum Length \* | Hardware Wallet Support |
+| :-: | :-------------------: | :---------------: | :---------------------: |
+|  0  | Restricted ASCII \*\* | 1232              | Yes                     |
+|  1  | UTF-8                 | 1232              | Blind sign only         |
+|  2  | UTF-8                 | 65535             | No                      |
+
+\* Combined length of the [message preamble](#message-preamble) and message body<br/>
+\*\* Those characters for which [`isprint(3)`](https://linux.die.net/man/3/isprint)
+returns true.  That is, `0x20..=0x7e`.
+
+Both the message encoding and maximum message length **MUST** be enforced by
+signer and verifier.
+
+Formats `0` and `1` are motivated by hardware wallet support where both RAM
+to store the payload and font character support are limited.
+
+This field **SHOULD NOT** be displayed to users
+
+## Signing
+
+Solana off-chain messages **MUST** only be signed using the ed25519 digital
+signature scheme. Before signing, the message **MUST** be strictly checked to
+conform to the associated preamble. The message body is then appended to the
+[message preamble](#message-preamble). Finally the result is ed25519 signed.
+
+## Verification
+Upon successful ed25519 verification of _all_ attached signatures, the message
+**MUST** be strictly checked to conform to the [message preamble](#message-preamble).
+A message that does not conform to its preamble is invalid, regardless of the
+validity of any signatures.
+
+## Envelope
+
+When passing around signed off-chain messages a common format is helpful. The
+recommended binary representation is as follows:
+
+| Field | Start offset | Length (bytes) | Description
+| :---- | :----------: | :------------: | :----------
+| Signature Count \* | 0x00 | 1 | Number of signatures. An 8-bit, unsigned integer
+| Signatures | 0x01 | `SIG_COUNT` * 64 | `SIG_COUNT` ed25519 signatures
+| Message Preamble | 0x01 + `SIG_COUNT` * 64 | `PREAMBLE_LEN` | The [message preamble](#message-preamble)
+| Message Body | 0x01 + `SIG_COUNT * 64 + `PREAMBLE_LEN` | `MESSAGE_LEN` | The message content
+
+The signature count **MUST** match the value of signers count from the message
+preamble.
+
+Signatures **MUST** be ordered to match their corresponding public keys as
+specified in the message preamble.
+
+## Runtime Considerations
+
+To prevent social attacks by which the signer is tricked into signing a transaction,
```

Comment:
shall we get rid of this paragraph given the #29807

Reply:
i think it's ok to leave it as a reference to what needed to be changed to accommodate offchain message signing. i'll add a link to your PR implementing it though.  good catch!

Reply:
https://github.com/solana-labs/solana/pull/26915/commits/c0857968ebf5a5be2dc24953389fccbe4e93217a#diff-e99c27ae1763d88fc8e45e699b94c3a9b4e8ccdab3a4a425d1d3db5012e70b7aR136-R138